### ISSUE TITLE
Room connecting (TwilioVideo.connect) is really slow

### ISSUE THREAD

**Issue by namtvz** [twilio/twilio-video-ios#31]
### Description

Room connecting with `TwilioVideo.connect(with: connectOptions, delegate: self)` is really slow. After this method is called, the `TVIRoomDelegate` callback `didConnect(to room: TVIRoom)` fires after 4 seconds (fastest, still bad but acceptable), or even 18 seconds (slowest that I met, not acceptable) later.

This is really not good for an instant VoIP app.

One note is I didn't meet this issue before (not sure from when) with same setup and test devices, versions.

### Steps to Reproduce

1. Add the SDK, set it up with own credentials
2. Call `TwilioVideo.connect(with: connectOptions, delegate: self)` to connect to a room (that was created in another side)
3. Wait for the `didConnect(to room: TVIRoom)` callback fires, log the different time

#### Code

```swift
TwilioVideo.connect(with: connectOptions, delegate: self)
```

#### Expected Behavior

`didConnect(to room: TVIRoom)` should fire after some seconds

#### Actual Behavior

It fires after 4 - 18 seconds

#### Reproduces How Often

Mostly around 6 - 8 seconds. Around 20% at 4 seconds. Around 5% at 18 seconds or so.

#### Logs

N/A

#### Video iOS SDK

2.5.3 via CocoaPods

#### Xcode

10.1

#### iOS Version

11.4.1, 12.1.1

#### iOS Device

iPhone 6, iPhone 7

**Comment by paynerc** [comment]
@namtvit07 

I'm sorry you are experiencing slow call setup times. We would like to hep you get to the bottom of this. I have a couple questions to help us get started.

1. What part of the world are you located in when joining the rooms? Wondering about network traversal, latency, and the like.
2. Do you have roomSids for calls where the setup duration is long? It would be helpful for us to be able to look at server side logs and see if we can see anything.
3. Most importantly, debug logs from the clients would be especially helpful. From there we could see what is happening during the call setup and see if we can pinpoint the cause.

To enable debug logging, add a call to:
```
    TwilioVideo.setLogLevel(.debug)
```

I typically do this in the AppDelegate's `application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?)` method.

If you are able to get debug logs and roomSids, please email them to support@twilio.com. Reference this Github issue in the message and also note that you are working with me and the ticket should find its way to me.

Looking forward to getting to the bottom of this with you,

Ryan

**Comment by namtvz** [comment]
@paynerc Thanks for your quick response.

I have sent the room sid and debug log to you. The Twilio ticket number is #1618521

1. What part of the world are you located in when joining the rooms? Wondering about network traversal, latency, and the like.

--> I'm in Ho Chi Minh City, Vietnam. My client in the US also meets slow connecting issue. I'm not 100% sure if it is also the connecting room issue like me (because I have not been able to debug on my client side yet) but I didn't find anything else.

In the debug log, the most spent time seems to be for DNS and TLS handshake. I'm not sure why they took much time. Is there any way to improve this? e.g: pre-populate something?

More information about the internet connection:

- Ping to the US is around 180ms
- Visiting websites from US is quick
- After connecting successfully, the call (both voice and video) is smooth

**Comment by paynerc** [comment]
@namtvit07,

Thank you for sending this over. I am investigating the logs for this room session and will get back to you shortly.

Thank you,

Ryan

**Comment by paynerc** [comment]
@namtvit07,

I have responded to the support ticket. Please look for the response there.

Thank you,

Ryan

**Comment by paynerc** [comment]
@namtvit07,

Just checking in to see if you have received my responses from the support ticket, #1618521.

Thank you,

Ryan

**Comment by namtvz** [comment]
@paynerc Sorry for the late response

I got the response on the ticket. However, the internet from Vietnam to US have met an issue recently, so I cannot assess the difference yet.

One quick question: per your suggestion, I would like to change `media_region`. In our current workflow, we create room on iOS side (with `TVIConnectOptions` and `TwilioVideo.connect`) with granted token from server side. I cannot find any way to set `media_region` on iOS SDK. Is this possible on iOS side or we need to change our flow (not a good idea in our scenario)?

**Comment by ceaglest** [comment]
Hi @namtvit07,

> I cannot find any way to set media_region on iOS SDK. Is this possible on iOS side or we need to change our flow (not a good idea in our scenario)?

Good question, I'll chime in on this point.

We know that using the REST API to create Rooms might not make sense for your use case, and for others. Because of this, we are planning on introducing SDK (iOS, Android, Browser) APIs to control signaling and media regions later in Q1. I'll share more information on this when I have it.

Best,
Chris

**Comment by seijik42** [comment]
Hello @ceaglest ,
Do you have plan to add media region option for iOS 2.x SDK?
We provide video call service in Japan, our users reported starting call is very slow.
Apparently SDK uses us1 media region even though our users are in Japan.
I found 3.x library has region option now, but 2.x doesn't.
I don't want to migrate to 3.x now, because it is still in beta. I don't want to use beta library in our production app.

**Comment by ceaglest** [comment]
Hi @seijik42,

> Do you have plan to add media region option for iOS 2.x SDK?

No, we aren't planning to add Client side APIs for region selection in 2.x.

> We provide video call service in Japan, our users reported starting call is very slow.
Apparently SDK uses us1 media region even though our users are in Japan.

Yes, the default media region for a Group Room is us1. You can create a Room with the jp1 media region today using the REST API and connect to it with 2.x Clients. We are also working on a "gll" media region, where media is automatically placed near to the Participants (This will require 3.x SDKs though).

One more thing to mention, we plan to offer a console drop down where you can change your default media region. This means that, for example, you could select jp1 as your default media region and have 2.x Clients connect to Ad-Hoc Rooms where the Media Server is in Japan.

> I found 3.x library has region option now, but 2.x doesn't.
I don't want to migrate to 3.x now, because it is still in beta. I don't want to use beta library in our production app.

We are extending our beta period into Q3, I can provide an update once we have a firm GA date.

Best,
Chris

**Comment by seijik42** [comment]
Hi @ceaglest Thank you for your response.

> we plan to offer a console drop down where you can change your default media region.

That's great! This is what we wanted because sometime we forget to provide media region option.

> We are extending our beta period into Q3, I can provide an update once we have a firm GA date.

This is good news too.
I'm looking forward to hearing update from you!
Best,
Seiji

**Comment by ceaglest** [comment]
Hi @seijik42,

We have now released [3.0.0-beta4](https://www.twilio.com/docs/video/changelog-twilio-video-ios-version-3x#300-beta4-september-6-2019) which has some connect performance improvements over earlier 3.0 betas. At this point the drop down for default media region is available in the console but behind a feature flag. If it's something you are interested in trying I could get you access for your account, along with documentation on selecting media regions for REST and Ad-Hoc Rooms.

We expect to GA 3.0 later in September, and are planning one final beta (3.0.0-beta5) before then.

Best,
Chris

**Comment by paynerc** [comment]
@seijik42,

`3.0` has been GA for some time and as we discussed prior should help you out with this issue. I am going to close this issue but if you are still having issues or questions, please feel free to reopen it.

Thank you,

Ryan
